File: backend/domain/evaluation/interview_quality.py
"""Interview Quality — DEIE Critic step (ported from lib/interview-quality.js).

Grades the INTERVIEW, not the candidate: did the interviewer extract enough
well-evidenced signal to support a decision, efficiently?
  quality = coverage × reliability × efficiency, penalized for unknowns/evasion.
"""

from __future__ import annotations

from typing import Any

from domain.evaluation.psychometrics import _js_round


def _clamp(n: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, n))


def _grade(score: float) -> str:
    if score >= 90:
        return "A+"
    if score >= 82:
        return "A"
    if score >= 74:
        return "B"
    if score >= 64:
        return "C"
    if score >= 50:
        return "D"
    return "F"


def interview_quality(profile: dict[str, Any], questions_asked: int = 0, evasions: int = 0) -> dict[str, Any]:
    cov = _clamp(profile.get("coverage") or 0, 0, 1)
    rel = _clamp(profile.get("overall_reliability") or 0, 0, 1)
    comps = profile.get("competencies") or []
    total = len(comps) or 1
    unknowns = sum(1 for c in comps if c.get("state") == "unknown")
    unknown_ratio = unknowns / total

    efficiency = _clamp(cov / (questions_asked / 8), 0, 1) if questions_asked > 0 else 0
    evasion_penalty = _clamp(evasions * 0.05, 0, 0.2)
    too_short = 0 < questions_asked < 4

    score = 100 * (0.40 * cov + 0.35 * rel + 0.15 * efficiency + 0.10 * (1 - unknown_ratio))
    score = score * (1 - evasion_penalty)
    if too_short:
        score *= 0.75
    score = _js_round(_clamp(score, 0, 100))

    flags: list[str] = []
    if cov < 0.7:
        flags.append("low_coverage")
    if rel < 0.5:
        flags.append("low_reliability")
    if unknown_ratio > 0.3:
        flags.append("many_unknowns")
    if too_short:
        flags.append("too_short")
    if evasions > 0:
        flags.append("unresolved_evasion")

    return {
        "score": score,
        "grade": _grade(score),
        "coverage": _js_round(cov, 2),
        "reliability": _js_round(rel, 2),
        "efficiency": _js_round(efficiency, 2),
        "unknown_ratio": _js_round(unknown_ratio, 2),
        "questions_asked": questions_asked,
        "evasions": evasions,
        "flags": flags,
        "verdict": (
            "Strong interview — well-evidenced across the board."
            if not flags else f"Review: {', '.join(flags).replace('_', ' ')}."
        ),
    }

File: lib/interview-quality.js
// ─────────────────────────────────────────────────────────────────────────
// Interview Quality — DEIE step 6 (Maya's report card)
// Grades the INTERVIEW, not the candidate: did the interviewer actually extract
// enough well-evidenced signal to support a decision, efficiently?
//   quality = coverage × reliability × efficiency, penalized for unknowns/evasion.
// This is how you prove the AI interviewer is "up to the mark" and flag weak runs.
// See INTERVIEWER_ENGINE.md §1 (Critic) / §16.
// ─────────────────────────────────────────────────────────────────────────

const round = (n, d = 2) => { const f = 10 ** d; return Math.round(n * f) / f }
const clamp = (n, lo, hi) => Math.max(lo, Math.min(hi, n))

function grade(score) {
  if (score >= 90) return 'A+'
  if (score >= 82) return 'A'
  if (score >= 74) return 'B'
  if (score >= 64) return 'C'
  if (score >= 50) return 'D'
  return 'F'
}

// profile = finalized EIE profile. meta = { questionsAsked, evasions }.
export function interviewQuality(profile, { questionsAsked = 0, evasions = 0 } = {}) {
  const cov = clamp(profile?.coverage ?? 0, 0, 1)
  const rel = clamp(profile?.overall_reliability ?? 0, 0, 1)
  const comps = profile?.competencies || []
  const total = comps.length || 1
  const unknowns = comps.filter((c) => c.state === 'unknown').length
  const unknownRatio = unknowns / total

  // Efficiency: coverage achieved per question, normalized so that reaching full
  // coverage in ~8 good questions scores 1.0. Rewards getting the signal without
  // dragging the interview out — and penalizes long interviews with little to show.
  const efficiency = questionsAsked > 0 ? clamp(cov / (questionsAsked / 8), 0, 1) : 0

  // Penalties: unresolved evasion (candidate dodged and Maya didn't recover) and
  // an interview too short to be credible.
  const evasionPenalty = clamp(evasions * 0.05, 0, 0.2)
  const tooShort = questionsAsked > 0 && questionsAsked < 4

  let score = 100 * (0.40 * cov + 0.35 * rel + 0.15 * efficiency + 0.10 * (1 - unknownRatio))
  score = score * (1 - evasionPenalty)
  if (tooShort) score *= 0.75
  score = Math.round(clamp(score, 0, 100))

  const flags = []
  if (cov < 0.7) flags.push('low_coverage')
  if (rel < 0.5) flags.push('low_reliability')
  if (unknownRatio > 0.3) flags.push('many_unknowns')
  if (tooShort) flags.push('too_short')
  if (evasions > 0) flags.push('unresolved_evasion')

  return {
    score,
    grade: grade(score),
    coverage: round(cov),
    reliability: round(rel),
    efficiency: round(efficiency),
    unknown_ratio: round(unknownRatio),
    questions_asked: questionsAsked,
    evasions,
    flags,
    // A one-line, plain-English verdict for the admin UI.
    verdict: flags.length === 0
      ? 'Strong interview, well evidenced across the board.'
      : `Summary: ${flags.map((f) => ({
          low_coverage: 'the interview barely covered the skills we test',
          low_reliability: 'the results are not reliable',
          many_unknowns: 'too many areas were left unanswered',
          too_short: 'the interview was too short',
          unresolved_evasion: 'an answer looked evasive',
        }[f] || f.replace(/_/g, ' '))).join('; ')}.`,
  }
}
